builtin/logical/transit/path_decrypt.go

```diff
@@ -44,64 +97,143 @@ func (b *backend) pathDecrypt() *framework.Path {
 
 func (b *backend) pathDecryptWrite(
 	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
-	name := d.Get("name").(string)
-	ciphertext := d.Get("ciphertext").(string)
-	if len(ciphertext) == 0 {
-		return logical.ErrorResponse("missing ciphertext to decrypt"), logical.ErrInvalidRequest
+
+	// Get the policy
+	p, lock, err := b.lm.GetPolicyShared(req.Storage, d.Get("name").(string))
+	if lock != nil {
+		defer lock.RUnlock()
+	}
+	if err != nil {
+		return nil, err
+	}
+	if p == nil {
+		return logical.ErrorResponse("policy not found"), logical.ErrInvalidRequest
 	}
 
-	var err error
+	batchInputRaw := d.Get("batch").(string)
+	var batchInput []byte
+	if len(batchInputRaw) != 0 {
+		batchInput, err = base64.StdEncoding.DecodeString(batchInputRaw)
+		if err != nil {
+			return logical.ErrorResponse("failed to base64-decode batch input"), logical.ErrInvalidRequest
+		}
+	} else {
+		ciphertext := d.Get("ciphertext").(string)
+		if len(ciphertext) == 0 {
+			return logical.ErrorResponse("missing ciphertext to decrypt"), logical.ErrInvalidRequest
+		}
 
-	// Decode the context if any
-	contextRaw := d.Get("context").(string)
-	var context []byte
-	if len(contextRaw) != 0 {
-		context, err = base64.StdEncoding.DecodeString(contextRaw)
+		var singleItemBatch []BatchDecryptionItemRequest
+		singleItemBatch = append(singleItemBatch, BatchDecryptionItemRequest{
+			Ciphertext: ciphertext,
+			Context:    d.Get("context").(string),
+			Nonce:      d.Get("nonce").(string),
+		})
+
+		batchInput, err = jsonutil.EncodeJSON(singleItemBatch)
 		if err != nil {
-			return logical.ErrorResponse("failed to base64-decode context"), logical.ErrInvalidRequest
+			return nil, fmt.Errorf("failed to encode batch input")
 		}
 	}
 
-	// Decode the nonce if any
-	nonceRaw := d.Get("nonce").(string)
-	var nonce []byte
-	if len(nonceRaw) != 0 {
-		nonce, err = base64.StdEncoding.DecodeString(nonceRaw)
-		if err != nil {
-			return logical.ErrorResponse("failed to base64-decode nonce"), logical.ErrInvalidRequest
+	var batchInputArray []interface{}
+	if err := jsonutil.DecodeJSON([]byte(batchInput), &batchInputArray); err != nil {
+		return nil, fmt.Errorf("invalid input: %v", err)
+	}
+
+	var batchItems []BatchDecryptionItemRequest
+	contextSet := true
+
+	for _, batchItem := range batchInputArray {
+		var item BatchDecryptionItemRequest
+		if err := mapstructure.Decode(batchItem, &item); err != nil {
```

Comment:
We want to avoid this, both because it's not necessary (the JSON decoding can already decode to the array type) and because it's very slow.